File: tokenmaxxing-macos/Sources/Tokenmaxxing/Creds.swift
import Foundation

struct ClaudeCredentials {
    var accessToken: String
    var refreshToken: String
    var expiresAtMs: Int64
    var subscriptionType: String
    var rateLimitTier: String
}

struct GrokCredentials {
    var accessToken: String
    var refreshToken: String
    var expiresAtMs: Int64
    var oidcIssuer: String
    var oidcClientId: String
    var email: String
    var tier: Int64
    var entryKey: String
}

/// Reads the same local credential files Claude Code, Grok, and opencode write.
/// Tokenmaxxing must run un-sandboxed for these home-directory reads to succeed.
enum Creds {
    static var home: URL { FileManager.default.homeDirectoryForCurrentUser }
    static var claudePath: URL { home.appending(path: ".claude/.credentials.json") }
    static var grokAuthPath: URL { home.appending(path: ".grok/auth.json") }
    static var grokSessionsPath: URL { home.appending(path: ".grok/sessions") }
    static var opencodeAuthPath: URL { home.appending(path: ".local/share/opencode/auth.json") }
    static var opencodeDbPath: URL { home.appending(path: ".local/share/opencode/opencode.db") }

    static func loadClaude() throws -> ClaudeCredentials {
        let data = try Data(contentsOf: claudePath)
        guard
            let root = try JSONSerialization.jsonObject(with: data) as? [String: Any],
            let oauth = root["claudeAiOauth"] as? [String: Any],
            let accessToken = oauth["accessToken"] as? String
        else {
            throw QuotaError.message("no claudeAiOauth block — run `claude` to sign in")
        }
        return ClaudeCredentials(
            accessToken: accessToken,
            refreshToken: oauth["refreshToken"] as? String ?? "",
            expiresAtMs: (oauth["expiresAt"] as? NSNumber)?.int64Value ?? 0,
            subscriptionType: oauth["subscriptionType"] as? String ?? "unknown",
            rateLimitTier: oauth["rateLimitTier"] as? String ?? ""
        )
    }

    /// Rewrite the OAuth tokens without disturbing sibling keys, atomically.
    static func writeBackClaude(accessToken: String, refreshToken: String, expiresAtMs: Int64) throws {
        let data = try Data(contentsOf: claudePath)
        guard var root = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              var oauth = root["claudeAiOauth"] as? [String: Any]
        else {
            throw QuotaError.message("credentials file missing claudeAiOauth")
        }
        oauth["accessToken"] = accessToken
        oauth["refreshToken"] = refreshToken
        oauth["expiresAt"] = expiresAtMs
        root["claudeAiOauth"] = oauth
        let out = try JSONSerialization.data(withJSONObject: root, options: [.prettyPrinted])
        try out.write(to: claudePath, options: [.atomic])
        try? FileManager.default.setAttributes([.posixPermissions: 0o600], ofItemAtPath: claudePath.path)
    }

    static func opencodeGoConfigured() -> Bool {
        guard let data = try? Data(contentsOf: opencodeAuthPath),
              let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return false }
        return root["opencode-go"] != nil
    }

    static func loadGrok() throws -> GrokCredentials {
        let data = try Data(contentsOf: grokAuthPath)
        guard let root = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw QuotaError.message("grok auth.json is not an object — run `grok login`")
        }
        let preferred = root.first { key, value in
            key.contains("auth.x.ai")
                && ((value as? [String: Any])?["key"] as? String).map { !$0.isEmpty } == true
        } ?? root.first { _, value in
            ((value as? [String: Any])?["key"] as? String).map { !$0.isEmpty } == true
        }
        guard let (entryKey, rawEntry) = preferred,
              let entry = rawEntry as? [String: Any],
              let accessToken = entry["key"] as? String
        else {
            throw QuotaError.message("no grok session — run `grok login`")
        }
        let expiresAt = entry["expires_at"] as? String ?? ""
        return GrokCredentials(
            accessToken: accessToken,
            refreshToken: entry["refresh_token"] as? String ?? "",
            expiresAtMs: parseExpiresMs(expiresAt),
            oidcIssuer: entry["oidc_issuer"] as? String ?? "https://auth.x.ai",
            oidcClientId: entry["oidc_client_id"] as? String ?? "",
            email: entry["email"] as? String ?? "",
            tier: jwtClaimInt64(accessToken, claim: "tier") ?? 0,
            entryKey: entryKey
        )
    }

    static func writeBackGrok(entryKey: String, accessToken: String, refreshToken: String, expiresAt: String) throws {
        let data = try Data(contentsOf: grokAuthPath)
        guard var root = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              var entry = root[entryKey] as? [String: Any]
        else {
            throw QuotaError.message("auth entry missing after refresh")
        }
        entry["key"] = accessToken
        entry["refresh_token"] = refreshToken
        entry["expires_at"] = expiresAt
        root[entryKey] = entry
        let out = try JSONSerialization.data(withJSONObject: root, options: [.prettyPrinted])
        try out.write(to: grokAuthPath, options: [.atomic])
        try? FileManager.default.setAttributes([.posixPermissions: 0o600], ofItemAtPath: grokAuthPath.path)
    }

    private static func parseExpiresMs(_ raw: String) -> Int64 {
        guard !raw.isEmpty else { return 0 }
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let date = formatter.date(from: raw) {
            return Int64(date.timeIntervalSince1970 * 1000)
        }
        formatter.formatOptions = [.withInternetDateTime]
        if let date = formatter.date(from: raw) {
            return Int64(date.timeIntervalSince1970 * 1000)
        }
        return 0
    }

    private static func jwtClaimInt64(_ token: String, claim: String) -> Int64? {
        let parts = token.split(separator: ".")
        guard parts.count >= 2 else { return nil }
        var payload = String(parts[1])
        let pad = (4 - payload.count % 4) % 4
        if pad > 0 { payload += String(repeating: "=", count: pad) }
        payload = payload.replacingOccurrences(of: "-", with: "+").replacingOccurrences(of: "_", with: "/")
        guard let data = Data(base64Encoded: payload),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return nil }
        if let n = json[claim] as? NSNumber { return n.int64Value }
        return nil
    }
}

enum QuotaError: Error, CustomStringConvertible {
    case message(String)
    var description: String {
        switch self {
        case .message(let value): value
        }
    }
}

File: ampere-kde/src/creds.rs
use std::path::PathBuf;

pub fn home() -> PathBuf {
    std::env::var_os("HOME")
        .map(PathBuf::from)
        .unwrap_or_default()
}

pub fn claude_credentials_path() -> PathBuf {
    home().join(".claude/.credentials.json")
}

pub fn opencode_auth_path() -> PathBuf {
    home().join(".local/share/opencode/auth.json")
}

pub fn opencode_db_path() -> PathBuf {
    home().join(".local/share/opencode/opencode.db")
}

#[derive(Clone, Debug)]
pub struct ClaudeCredentials {
    pub access_token: String,
    pub refresh_token: String,
    pub expires_at_ms: i64,
    pub subscription_type: String,
    pub rate_limit_tier: String,
}

pub fn load_claude() -> Result<ClaudeCredentials, String> {
    let path = claude_credentials_path();
    let raw = std::fs::read_to_string(&path)
        .map_err(|e| format!("cannot read {}: {e}", path.display()))?;
    let json: serde_json::Value =
        serde_json::from_str(&raw).map_err(|e| format!("bad JSON in credentials: {e}"))?;
    let oauth = json
        .get("claudeAiOauth")
        .ok_or("no claudeAiOauth block — run `claude` to sign in")?;

    let access_token = oauth
        .get("accessToken")
        .and_then(|v| v.as_str())
        .ok_or("no accessToken")?
        .to_string();
    let refresh_token = oauth
        .get("refreshToken")
        .and_then(|v| v.as_str())
        .unwrap_or_default()
        .to_string();
    let expires_at_ms = oauth
        .get("expiresAt")
        .and_then(serde_json::Value::as_i64)
        .unwrap_or(0);
    let subscription_type = oauth
        .get("subscriptionType")
        .and_then(|v| v.as_str())
        .unwrap_or("unknown")
        .to_string();
    let rate_limit_tier = oauth
        .get("rateLimitTier")
        .and_then(|v| v.as_str())
        .unwrap_or_default()
        .to_string();

    Ok(ClaudeCredentials {
        access_token,
        refresh_token,
        expires_at_ms,
        subscription_type,
        rate_limit_tier,
    })
}

/// True when the opencode-go subscription is configured on this machine.
pub fn opencode_go_configured() -> bool {
    let Ok(raw) = std::fs::read_to_string(opencode_auth_path()) else {
        return false;
    };
    serde_json::from_str::<serde_json::Value>(&raw)
        .ok()
        .and_then(|v| v.get("opencode-go").cloned())
        .is_some()
}
